tests/unit/schema/test_api.py

```diff
@@ -2602,3 +2607,374 @@ def test_has_iam_mixin():
+    api_schema = api.API.build(fd, "google.example.v1beta1", opts=cli_options)
+    assert api_schema.all_method_settings == {
+        "google.example.v1beta1.SomeExample.Example1": client_pb2.MethodSettings(
+            selector="google.example.v1beta1.SomeExample.Example1",
+            auto_populated_fields=["squid", "mollusc"],
+            long_running=client_pb2.MethodSettings.LongRunning(),
+        )
+    }
+
+
+def test_method_settings_duplicate_selector_raises_error():
+    """
+    Test that `MethodSettingsError` is raised when there are duplicate selectors in
+    `client_pb2.MethodSettings`.
+    """
+    fd = get_file_descriptor_proto_for_method_settings_tests()
+    api_schema = api.API.build(fd, "google.example.v1beta1")
+    methodsettings = [
+        client_pb2.MethodSettings(
+            selector="google.example.v1beta1.SomeExample.Example1",
+        ),
+        client_pb2.MethodSettings(
+            selector="google.example.v1beta1.SomeExample.Example1",
+        ),
+    ]
+    with pytest.raises(
+        api.MethodSettingsError, match="(?i)duplicate selector"
+    ):
+        api_schema.enforce_valid_method_settings(methodsettings)
+
+
+def test_method_settings_invalid_selector_raises_error():
+    """
+    Test that `MethodSettingsError` when `client_pb2.MethodSettings.selector`
+    cannot be mapped to a method in the API.
+    """
+    fd = get_file_descriptor_proto_for_method_settings_tests()
+    api_schema = api.API.build(fd, "google.example.v1beta1")
+    methodsettings = [
+        client_pb2.MethodSettings(
+            selector="google.example.v1beta1.DoesNotExist.Example1",
+        ),
+    ]
+    with pytest.raises(
+        api.MethodSettingsError, match="(?i)method was not found"
+    ):
+        api_schema.enforce_valid_method_settings(methodsettings)
+
+
+def test_method_settings_unsupported_auto_populated_field_type_raises_error():
+    """
+    Test that `MethodSettingsError` is raised when a field in
+    `client_pb2.MethodSettings.auto_populated_fields` is not of type string.
+    """
+    squid = make_field_pb2(name="squid", type="TYPE_INT32", number=1)
+    fd = get_file_descriptor_proto_for_method_settings_tests(fields=[squid])
+    api_schema = api.API.build(fd, "google.example.v1beta1")
+    methodsettings = [
+        client_pb2.MethodSettings(
+            selector="google.example.v1beta1.SomeExample.Example1",
+            auto_populated_fields=["squid"],
+        ),
+    ]
+    with pytest.raises(api.MethodSettingsError, match="(?i)type string"):
+        api_schema.enforce_valid_method_settings(methodsettings)
+
+
+def test_method_settings_auto_populated_field_not_found_raises_error():
+    """
+    Test that `MethodSettingsError` is raised when a field in
+    `client_pb2.MethodSettings.auto_populated_fields` is not found in the top-level
+    request message of the selector.
+    """
+    fd = get_file_descriptor_proto_for_method_settings_tests()
```

Comment:
Consider populating some other field of the message.

Also, is it simple to test when the field in `auto_populated_fields` is nested one level in ? (I.e., it still exists but not at the top-level, so we still get the exception)

Reply:
Done in https://github.com/googleapis/gapic-generator-python/pull/1975/commits/1ab06eae2427367f3bdf5f7ce5dc659345265e2b